Trait formatting (#101)
WordPress/wpcs-docs: wordpress-coding-standards/php.md

);
```

> ### Trait Formatting

Reviewer: ```suggestion
### Trait Use statements
```

Trait `use` statements should be at the top of a class and should have exactly one blank line before the first `use` statement, and at least one blank line after. The only exception is when the class only contains the trait `use` statement, in which case the blank line after may be omitted.

> The following code examples cover the rule about formatting traits regarding things like spacing, grouping and indentation.

Reviewer: ```suggestion
The following code examples show the formatting requirements for trait `use` statements regarding things like spacing, grouping and indentation.
```

```php
// Correct.
class Foo {

>     // Comments above a use statement will be regarded as documentation for the statement.

Reviewer: I think this comment line can be omitted.

    use Bar_Trait;
    use Foo_Trait,
        Bazinga_Trait {
        Bar_Trait::method_name insteadof Bar_Trait;
        Bazinga_Trait::method_name as bazinga_method;
    }
    use Loopy_Trait { eat as protected; }

    public $baz = true;

    ...
}

class Foo {

    use Bar_Trait;

> }

Reviewer: I think this code sample can be omitted.

// Incorrect.
class Foo {
    use       Bar_Trait; // No blank line before class declaration, multiple spaces after the use statement.

    use Foo_Trait,   Bazinga_Trait{Bar_Trait::method_name    insteadof     Foo_Trait; Bazinga_Trait::method_name // Multiple spaces when importing traits, no new line after opening brace
        as     bazinga_method; // Multiple spaces, should be on the same line as the method it's replacing.
            }; // Wrongly indented brace.

>     public $baz = true; // Missing blank line after trait import.

Reviewer: This code is hard to read as it is (which is exactly the point) and the comments make it even harder.

Maybe list out the things which are wrong in a multi-line `/* .. */` comment above the `use` statement ?

class Example_Class_Extended { [...] }
```

### Trait Use Statements

> Trait `use` statements should be at the top of a class and should have exactly one blank line before the first `use` statement, and at least one blank line after. The only exception is when the class only contains the trait `use` statement, in which case the blank line after may be omitted.

Reviewer: > Going from potentially an infinite amount of blank lines after, to potentially zero would seem to make it harder to enforce than necessary. Can we change the "at least one blank line after" to [exactly] "one blank line after"? That way, it's either one line after, or zero lines after if that's the only thing in the class.

This was a very deliberate choice of wording, both for the class only containing trait `use` as well as the class with more content. We use the same wording in multiple other places too.

In this case, whatever _follows_ the trait `use` statement determines how many blank lines _between_ the `use` and the next thing are allowed. The trait `use` "blank lines after" only determines that there should be _at least_ one blank line.

In more general terms: strict rules for blank lines before something, loose rules for blank lines after.

If we don't do it this way, it can very easily lead to fixer conflicts, which is something we want to avoid.

As for the "class only containing trait `use` may omit the blank line after" part, again, this is largely to prevent fixer conflicts as generally speaking, there should be no stray blank lines at the end of a class and sniffs checking for that normally don't care what is directly above the closing brace, as long as it isn't a blank line.

> Trait `use` statements should be at the top of a class and should have exactly one blank line before the first `use` statement, and at least one blank line after. The only exception is when the class only contains the trait `use` statement, in which case the blank line after may be omitted.

Reviewer: @GaryJones Does it make more sense when you read it with that context in mind ?

> Trait `use` statements should be at the top of a class and should have exactly one blank line before the first `use` statement, and at least one blank line after. The only exception is when the class only contains the trait `use` statement, in which case the blank line after may be omitted.

Reviewer: @dingo-d Looks like @GaryJones' suggestion to remove ambiguity hasn't been applied yet.

The following code examples show the formatting requirements for trait `use` statements regarding things like spacing, grouping and indentation.

```php
// Correct.
class Foo {

    use Bar_Trait;
    use Foo_Trait,
        Bazinga_Trait {
        Bar_Trait::method_name insteadof Bar_Trait;
        Bazinga_Trait::method_name as bazinga_method;
    }
    use Loopy_Trait { eat as protected; }

    public $baz = true;

    ...
}

// Incorrect.
class Foo {

>     /**

Reviewer: I concur. Docblock is the wrong format. Either multiple `//` or a multi-line inline comment `/* ... */` (note: single `*`).

>      * No blank line before class declaration, multiple spaces after the use statement.

Reviewer: + end of sentence: "use statement" => "`use` keyword"
